talk/owt/sdk/p2p/p2ppeerconnectionchannel.cc

```diff
@@ -163,6 +163,144 @@ void P2PPeerConnectionChannel::ClearPendingStreams() {
+                            "The stream is null."));
+          on_failure(std::move(e));
+        }
+    }
+  });
+
+  std::vector<std::string> stream_labels_to_prune;
+  {
+    std::lock_guard<std::mutex> lock(published_streams_mutex_);
+    for (const auto &stream_label : stream_labels) {
+      if (published_streams_.find(stream_label) != published_streams_.end() ||
+          publishing_streams_.find(stream_label) != publishing_streams_.end()) {
+        // Prune the streams that are already published
+        streams.erase(std::remove_if(streams.begin(), streams.end(), [stream_label](std::shared_ptr<LocalStream> stream) {
+          return stream->MediaStream()->id() == stream_label;
+        }), streams.end());
+        stream_labels_to_prune.push_back(stream_label);
+
+        if (on_failure) {
+          std::unique_ptr<Exception> e(
+              new Exception(ExceptionType::kP2PClientInvalidArgument,
+                            "The stream is already published."));
+          on_failure(std::move(e));
+        }
+      }
+    }
+  }
+
+  // Prune stream labels of already published/publishing streams
+  stream_labels.erase(std::remove_if(stream_labels.begin(), stream_labels.end(), [&stream_labels_to_prune](std::string stream_label) {
+      return std::find(stream_labels_to_prune.begin(), stream_labels_to_prune.end(), stream_label) != stream_labels_to_prune.end();
+    }), stream_labels.end());
+
+  // Prune null streams
+  streams.erase(std::remove_if(streams.begin(), streams.end(), [](std::shared_ptr<LocalStream> stream) {
+    return !stream;
+  }), streams.end());
+
+
+  for (auto stream : streams) {
+    RTC_LOG(LS_INFO) << "Publishing a local stream.";
+
+    rtc::scoped_refptr<webrtc::MediaStreamInterface> media_stream =
+        stream->MediaStream();
+    // Calling [stream|track]->id() runs on signaling_thread, so call outside of locks
+    std::string stream_id = media_stream->id();
+    std::pair<std::string, std::string> stream_track_info;
+    for (const auto& track : media_stream->GetAudioTracks()) {
+      std::string track_id = track->id();
+      {
+        std::lock_guard<std::mutex> lock(local_stream_tracks_info_mutex_);
+        if (local_stream_tracks_info_.find(track_id) ==
+            local_stream_tracks_info_.end()) {
+          stream_track_info = std::make_pair(track_id, stream_id);
+          local_stream_tracks_info_.insert(stream_track_info);
+        }
+      }
+    }
+    for (const auto& track : media_stream->GetVideoTracks()) {
+      std::string track_id = track->id();
+      {
+        std::lock_guard<std::mutex> lock(local_stream_tracks_info_mutex_);
+        if (local_stream_tracks_info_.find(track_id) ==
+            local_stream_tracks_info_.end()) {
+          stream_track_info = std::make_pair(track_id, stream_id);
+          local_stream_tracks_info_.insert(stream_track_info);
+        }
+      }
+    }
+  }
+
+  // Send chat-closed to workaround known browser bugs, together with
```

Comment:
this won't have a functional impact, but just so it's easier to compare to `Publish()`, let's move it above the for loop